### PR TITLE
Longhorn Online Rebuilding

### PR DESCRIPTION
**Pull request by shuo-wu**
#### Which issue(s) this PR fixes:

Issue longhorn/longhorn#7199

### CODE REVIEW

File: controller/engine_controller.go [hunk 1]
```diff
@@ -1948,6 +1948,44 @@ func (ec *EngineController) startRebuilding(e *longhorn.Engine, replicaName, add
 				"Start rebuilding replica %v with Address %v for normal engine %v and volume %v", replicaName, addr, e.Name, e.Spec.VolumeName)
 			err = engineClientProxy.ReplicaAdd(e, replicaName, replicaURL, false, fastReplicaRebuild, localSync, fileSyncHTTPClientTimeout, grpcTimeoutSeconds)
 		}
+
+		// For v2 engine, the rebuilding is an async call. We need to wait for the rebuilding start then complete here
```
Comment (derekbit) [inline controller/engine_controller.go:1952]: Is the resource reconciliation stuck here until the rebuild is complete? If the controller has 5 workers and all are waiting for v2 rebuilding, does this mean no workers are available to reconcile engine resources? Should we use a goroutine for the asynchronous call?

File: controller/engine_controller.go [hunk 2]
```diff
@@ -1948,6 +1948,44 @@ func (ec *EngineController) startRebuilding(e *longhorn.Engine, replicaName, add
...
+		if err == nil && types.IsDataEngineV2(e.Spec.DataEngine) {
+			ticker := time.NewTicker(EnginePollInterval)
+			defer ticker.Stop()
+			timer := time.NewTimer(time.Duration(grpcTimeoutSeconds) * time.Second)
+			defer timer.Stop()
+			finished := false
+			for {
+				select {
```
Comment (derekbit) [inline controller/engine_controller.go:1960]: Suggest abstracting the long codes for v2 engine to a function.

**Review by derekbit — COMMENTED** [review]
LGTM. Only one comment for readability.

**Review by derekbit — DISMISSED** [review]
LGTM

File: scheduler/replica_scheduler.go [hunk 3]
```diff
@@ -654,6 +658,10 @@ func (rcs *ReplicaScheduler) RequireNewReplica(replicas map[string]*longhorn.Rep
 		return 0
 	}
 
+	if types.IsDataEngineV2(volume.Spec.DataEngine) {
+		return 0
+	}
+
```
Comment (derekbit) [inline scheduler/replica_scheduler.go:664]: I missed this check.
What's the purpose of it?

Comment (shuo-wu) [reply scheduler/replica_scheduler.go:664]: This means, we should wait 0s for replenishing a new replica for v2 volumes.

**Review by derekbit — APPROVED** [review]

**Comment by derekbit** [comment]
@mergify backport v1.7.x

**Pull request merged**